### PR TITLE
Start logging events with our structured logger

### PR DESCRIPTION
**Pull request by rmohr**
**What this PR does / why we need it**:

So far we missed some events in the logs if we did not also write an extra log message. Start logging each even in the log stream of each component. 

Grepping a log for `"event"` will reveal all event entries.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
```release-note
Start writing events to our structured logger
```

### CODE REVIEW

**Comment by rmohr** [comment]
/cc @davidvossel 

FYI

I did not yet check how much duplications we will get with that.

**Review by davidvossel — COMMENTED** [review]
@rmohr  unless we go through and look at every place an event is sent and remove the duplicate logs, this will start to add double log entries for all the places we both send and event and added a matching log entry. 

for example, there are a lot of places in the code that couple an event with a log msg. Now we'll get two similar log messages for each event unless this is cleaned up

**Comment by rmohr** [comment]
> @rmohr unless we go through and look at every place an event is sent and remove the duplicate logs, this will start to add double log entries for all the places we both send and event and added a matching log entry.
> 
> for example, there are a lot of places in the code that couple an event with a log msg. Now we'll get two similar log messages for each event unless this is cleaned up

I will have a look, but I am not convinced that it is necessary to de-duplicate it. I marked the log string to make it clear that it is an event which gets logged. That is happening asynchronously.

**Comment by ezrasilvera** [comment]
@rmohr So, if I understand it correctly all events will now be logged without any filtering ability. Is that correct? Is that the desired behavior (i.e., we must have logs of all events) ?
Do we expect any performance impact of such additional logging ?

**Comment by rmohr** [comment]
> @rmohr So, if I understand it correctly all events will now be logged without any filtering ability. Is that correct? Is that the desired behavior (i.e., we must have logs of all events) ?

Yes. I think it is very helpful to have the ability to find reported events also in the component logs.

> Do we expect any performance impact of such additional logging ?

Not really, it should be a small subset compared to the logs which we anyway log on every sync operation

**Comment by ezrasilvera** [comment]
> > Do we expect any performance impact of such additional logging ?
> 
> Not really, it should be a small subset compared to the logs which we anyway log on every sync operation

Sounds good.  If possible it would be good to have a possibility to disable the auto event logs for modules/components . Just in case we may have high frequency events in the future. 
If we are sure currently we don't expect performance or file-size issues, I guess we can postpone it for the future (as long as there is a way to do it) .   Maybe we can add an event-type of "noAudoLoggerSink" and filer it out...

**Comment by ezrasilvera** [comment]
> > @rmohr unless we go through and look at every place an event is sent and remove the duplicate logs, this will start to add double log entries for all the places we both send and event and added a matching log entry.
> > for example, there are a lot of places in the code that couple an event with a log msg. Now we'll get two similar log messages for each event unless this is cleaned up
> 
> I will have a look, but I am not convinced that it is necessary to de-duplicate it. I marked the log string to make it clear that it is an event which gets logged. That is happening asynchronously. Also the focus of a log message can be different. Different log level, more or less verbose, ...

I think for some cases we do want to avoid such duplications especially as they probably going to appear in different places as the event log is async (right ?) 
Personally I think it will be good enough to explicitly open PRs for such cases as someone encounter them, not sure we really need to manually check this as part this PR (it will also add files not directly related to the logging change to the PR)

**Comment by davidvossel** [comment]
>Personally I think it will be good enough to explicitly open PRs for such cases as someone encounter them, not sure we really need to manually check this as part this PR (it will also add files not directly related to the logging change to the PR)

@rmohr  how would you feel about just checking for duplications in the migration, vmi, vm, and virt-hander vm controllers? I think the migration controller is the worst offender here. Nearly every event is coupled with a log message (my doing probably) and there are a few other occurrences in those other controllers.

**Review by davidvossel — COMMENTED** [review]
/lgtm
/approve
/hold

@rmohr i pushed a change that removes duplicated logging due to the new event logger change. see if it makes sense to you and remove the hold if you're comfortable with it.

File: staging/src/kubevirt.io/client-go/log/log.go [hunk 1]
```diff
@@ -433,3 +433,10 @@ func LogQemuLogLine(logger *FilteredLogger, line string) {
 		"msg", line,
 	)
 }
+func EventLoggerSink(event *v1.Event) {
+	if event.Type == v1.EventTypeWarning {
+		DefaultLogger().ObjectRef(&event.InvolvedObject).With("event", event.Type).Reason(fmt.Errorf(event.Reason)).Error(event.Message)
+	} else {
+		DefaultLogger().ObjectRef(&event.InvolvedObject).With("event", event.Type).Reason(fmt.Errorf(event.Reason)).Info(event.Message)
+	}
+}
```
Comment (davidvossel) [inline staging/src/kubevirt.io/client-go/log/log.go:442]: the log message says it originates from log.go line 440 for all log messages. we need to have an offset that reflects the actual line the log line originates from. Below is an example of what i mean. 

```{"component":"virt-controller","event":"Normal","kind":"VirtualMachineInstance","level":"info","msg":"Created virtual machine pod virt-launcher-vmi-ephemeral-x4ch5","name":"vmi-ephemeral","namespace":"default","pos":"log.go:440","reason":"SuccessfulCreate","timestamp":"2021-08-20T19:13:27.001810Z","uid":"5e93fea8-4c9e-452a-9daa-b8d2230b965e"}```

**Review by davidvossel — CHANGES_REQUESTED** [review]
/lgtm cancel

log lines in log output is not accurate

File: staging/src/kubevirt.io/client-go/log/log.go (hunk 1 above)
Comment (davidvossel) [reply staging/src/kubevirt.io/client-go/log/log.go:442]: @rmohr any ideas here?

Comment (rmohr) [reply staging/src/kubevirt.io/client-go/log/log.go:442]: We introduced a new layer from the stack perspective. We have the depth in some helpers encoded. For instance

```
func (l FilteredLogger) msgf(msg string, args ...interface{}) {
	l.log(3, "msg", fmt.Sprintf(msg, args...))
}
```

where we have `3`. A small extension to `log.go` where one can set the offset should be good enough.

**Comment by rmohr** [comment]
@davidvossel do you think you can find the time to pick this up again?

**Comment by davidvossel** [comment]
> @davidvossel do you think you can find the time to pick this up again?

It's unlikely I'll give this attention anytime soon.

**Pull request closed without merging**